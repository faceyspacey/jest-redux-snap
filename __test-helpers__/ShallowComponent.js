import React from 'react'


const ShallowComponent = ({ className }) =>
  <div id='shallowComponent' className={className} />

export default ShallowComponent
