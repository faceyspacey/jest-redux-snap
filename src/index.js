export { default as snap } from './snap'
export { default as createApp } from './createApp'
