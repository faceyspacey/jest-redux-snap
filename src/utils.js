// eg: MyComponent
export const isClass = Component => typeof Component === 'function' && !isElement(Component) && !isInstance(Component)

// eg: renderer.create(<MyComponent />).toJSON()
export const isTree = Component => Component && Component.$$typeof && Component.$$typeof.toString() === 'Symbol(react.test.json)'

// eg: <MyComponent />
export const isElement = Component => Component && Component.$$typeof

// eg: renderer.create(<MyComponent />)
export const isInstance = Component => Component && Component.toJSON


export const isStore = arg => typeof arg === 'object' && arg.getState
