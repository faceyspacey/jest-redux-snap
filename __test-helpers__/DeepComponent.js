import React from 'react'
import ShallowComponent from './ShallowComponent'


const DeepComponent = ({ title = 'A TITLE', className = 'myClass' }) =>
  <div>
    <span>{title}</span>
    <ShallowComponent className={className} />
  </div>

export default DeepComponent
