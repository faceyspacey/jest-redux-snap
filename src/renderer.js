import { isInstance } from './utils'


export function DeepRenderer(instance, tree) {
  this._instance = instance
  this._tree = tree
}

DeepRenderer.prototype = {
  // calling tree() insures having up to date instance in response to state changes
  tree() {
    return isInstance(this._instance) ? this._instance.toJSON() : this._tree
  },

  // calling snap will of course also use most up to date instance
  snap() {
    expect(this.tree()).toMatchSnapshot()
    return this
  },

  // this._instance can be passed to snap(renderer.instance()) -- this is as opposed to doing this.snap()
  component() {
    return this._instance
  },
}


export function ShallowRenderer(renderer, element, tree) {
  this._renderer = renderer
  this._element = element
  this._tree = tree
}

ShallowRenderer.prototype = {
  tree() {
    return this._element ? this._renderer.getRenderOutput() : this._tree
  },
  snap() {
    expect(this.tree()).toMatchSnapshot()
    return this
  },
  component() {
    return this._element
  },
}
