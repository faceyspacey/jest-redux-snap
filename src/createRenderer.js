import React from 'react'
import renderer from 'react-test-renderer'
import ReactTestUtils from 'react-addons-test-utils'

import { isClass, isTree, isElement, isInstance } from './utils'
import { DeepRenderer, ShallowRenderer } from './renderer'


export default function createRenderer(target, props, deep = true) {
  if (typeof props === 'boolean') {
    deep = props
    props = undefined
  }

  if (!deep) {
    return _createShallowRenderer(target, props)
  }

  let tree
  let instance

  if (isClass(target)) {
    instance = renderer.create(<target {...props} />)
  }
  else if (isTree(target)) {
    tree = target
  }
  else if (isElement(target)) {
    instance = renderer.create(target)
  }
  else if (isInstance(target)) {
    instance = target
  }
  else {
    tree = target
  }

  return new DeepRenderer(instance, tree)
}


export function _createShallowRenderer(target, props) {
  const renderer = ReactTestUtils.createRenderer()

  let tree
  let element

  if (isClass(target)) {
    element = <target {...props} />
    renderer.render(element)
  }
  else if (isTree(target)) {
    tree = target
  }
  else if (isElement(target)) {
    element = target
    renderer.render(element)
  }
  else {
    tree = target
  }

  return new ShallowRenderer(renderer, element, tree)
}


// ReactTestUtils does not create component instances that play nice with Jest mocks.
// A mocked component passed to ReactTestUtils.createRenderer().render(component)
// will expect there to be no props passed, which defeats its purpose.

const originalError = global.console.error.bind(global.console)

global.console.error = (...args) => {
  const msg = args[0]
  if (msg && msg.indexOf('Unknown prop') > -1) {
    return
  }

  return originalError(...args)
}
