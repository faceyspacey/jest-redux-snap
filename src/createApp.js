import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'

import createRenderer from './createRenderer'
import snap from './snap'


export default (store, ReactComponentClass, mapStateToProps) => {
  let wrapper

  store.subscribe(() => {
    wrapper.forceUpdate()
  })

  const element =
    <Provider store={store}>
      <ReactiveWrapper
        ref={wr => wrapper = wr}
        store={store}
        Component={ReactComponentClass}
        mapStateToProps={mapStateToProps}
      />
    </Provider>

  const instance = renderer.create(element)
  const app = createRenderer(instance)

  addMethods(app, store, element)

  return app
}


class ReactiveWrapper extends React.Component {
  render() {
    const { store, Component, mapStateToProps } = this.props

    const props = typeof mapStateToProps === 'function'
      ? mapStateToProps(store.getState(), store.dispatch)
      : mapStateToProps // object or undefined

    return <Component {...props} />
  }
}


const addMethods = (app, store, element) => {
  app.element = () => element
  app.story = () => element

  app.getState = () => store.getState()
  app.dispatch = action => store.dispatch(action)

  app.snapAction = (action) => {
    if (typeof action !== 'function') {
      snap(action)
    }

    store.dispatch(action)
    snap(store.getState())
    app.snap()
  }

  app.snapState = () => snap(app.getState())
}
