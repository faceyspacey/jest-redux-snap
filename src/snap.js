import createRenderer from './createRenderer'

export default function snap(target, props, deep = true) {
  return createRenderer(target, props, deep).snap()
}
