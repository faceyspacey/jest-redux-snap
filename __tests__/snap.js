import React from 'react'
import renderer from 'react-test-renderer'

import DeepComponent from '../__test-helpers__/DeepComponent'
import ShallowComponent from '../__test-helpers__/ShallowComponent'

import { snap } from '../src'


describe('snap(target, props) - deep', () => {
  const Target = DeepComponent

  it('target == class', () => {
    const app = snap(Target, { className: 'foo' })

    console.log(app.tree())
    console.log(app.component())
  })

  it('target == tree', () => {
    const target = renderer.create(<Target className='baz' />).toJSON()
    const app = snap(target)

    console.log(app.tree())

    expect(app.component()).not.toBeDefined()
  })

  it('target == element', () => {
    const target = <Target className='bar' />
    const app = snap(target)

    console.log(app.tree())
  })

  it('target == instance', () => {
    const target = renderer.create(<Target className='baz' />)
    const app = snap(target)

    console.log(app.tree())

    expect(app.component()).toEqual(target)
  })

  it('target == any', () => {
    const target = { plain: 'object' }
    const app = snap(target)

    console.log(app.component())
    console.log(app.tree())

    expect(app.component()).not.toBeDefined()
  })
})


describe('snap(target, props, false) - shallow', () => {
  const Target = ShallowComponent

  it('target == class', () => {
    const app = snap(Target, { className: 'foo' }, false)

    console.log(app.tree())
    expect(app.component()).toEqual(<Target className='foo' />)
  })

  it('target == tree', () => {
    const target = renderer.create(<Target className='baz' />).toJSON()
    const app = snap(target, false)

    console.log(app.tree())

    expect(app.component()).not.toBeDefined()
  })

  it('target == element', () => {
    const target = <Target className='bar' />
    const app = snap(target, false)

    console.log(app.tree())
    expect(app.component()).toEqual(target)
})

  it('target == any', () => {
    const target = { plain: 'object' }
    const app = snap(target, false)

    console.log(app.component())
    console.log(app.tree())

    expect(app.component()).not.toBeDefined()
  })
})
