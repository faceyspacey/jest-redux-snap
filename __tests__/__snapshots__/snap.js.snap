// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snap(target, props) - deep target == any 1`] = `
Object {
  "plain": "object",
}
`;

exports[`snap(target, props) - deep target == class 1`] = `
<div>
  <span>
    A TITLE
  </span>
  <div
    className="foo"
    id="shallowComponent"
  />
</div>
`;

exports[`snap(target, props) - deep target == element 1`] = `
<div>
  <span>
    A TITLE
  </span>
  <div
    className="bar"
    id="shallowComponent"
  />
</div>
`;

exports[`snap(target, props) - deep target == instance 1`] = `
<div>
  <span>
    A TITLE
  </span>
  <div
    className="baz"
    id="shallowComponent"
  />
</div>
`;

exports[`snap(target, props) - deep target == tree 1`] = `
<div>
  <span>
    A TITLE
  </span>
  <div
    className="baz"
    id="shallowComponent"
  />
</div>
`;

exports[`snap(target, props, false) - shallow target == any 1`] = `
Object {
  "plain": "object",
}
`;

exports[`snap(target, props, false) - shallow target == class 1`] = `
<div
  className="foo"
  id="shallowComponent"
/>
`;

exports[`snap(target, props, false) - shallow target == element 1`] = `
<div
  className="bar"
  id="shallowComponent"
/>
`;

exports[`snap(target, props, false) - shallow target == tree 1`] = `
<div
  className="baz"
  id="shallowComponent"
/>
`;
